### PR TITLE
Add XCTS system equations

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

This PR adds the XCTS system equations. An upcoming PR will add the XCTS system, fluxes and sources that use these functions and that will also call into the `Poisson` and the `Elasticity` code to for the remaining operations. This is because the XCTS equations are essentially two Poisson equations and an elasticity equation on a curved geometry, coupled through nonlinear sources.

Dependencies:

- [x] #2876 

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

**Comment by nilsvu** [comment]
@markscheel @haraldp271 @geoffrey4444 @nilsdeppe Here, at last, are the XCTS equations. Would two of you be willing to review the code? The PR is, of course, quite big and almost exclusively tensor-math 🤯😄

**Comment by nilsdeppe** [comment]
This seems like it should be 2-3 PRs... There are several "short" commits (200-400 lines) and then the giant XCTS one. Does it really all need to be one monster PR?

**Comment by nilsvu** [comment]
Sure, I can factor the Poisson and Elasticity commits into separate PRs. I kept them here because they provide terms for the XCTS equations and it might actually be better to review them in the same context. It's a lot of math either way ^^ I'll do what reviewers prefer

**Comment by nilsdeppe** [comment]
So I at least would be able to review the non-XCTS part of this PR fairly "casually" and not feel like I'm committing so something huge (so on the time scale of a few days, maybe faster)

**Comment by nilsvu** [comment]
Ok here's the first 3 commits as a separate PR: #2674

**Comment by nilsvu** [comment]
Also factored out the fourth commit here: #2675

File: src/Elliptic/Systems/Elasticity/Equations.cpp
Comment (haraldp271) [inline src/Elliptic/Systems/Elasticity/Equations.cpp]: These are a lot of duplicated multiplications by 0.5.  Can line 91 be written as 

`const auto co_displacement_over2 = 0.5*raise_or_lower_index(displacement, metric);
`

File: src/Elliptic/Systems/Poisson/Equations.cpp
Comment (haraldp271) [inline src/Elliptic/Systems/Poisson/Equations.cpp]: I find the differing syntax for get surprising.  Why not `source_for_field->get()` ?

Probably a design decision already discussed elsewhere, if so, no long explanation necessary.

File: src/Elliptic/Systems/Poisson/Equations.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Poisson/Equations.hpp]: It seems that `Poisson` refers to a **scalar** Poisson equation only.  That's fine, but please make sure this is mentioned near the start of the primary Poisson documentation.

File: src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp]:
```suggestion
 * \f$w_i=\partial_i\left(\alpha\psi\right)\f$.  Please note that $B_{ij}$ includes its  
 * trace-part, i.e. it differs from the longitudinal operator $(L\beta)_{ij}$ from 
 * the NR literature. 
 *
 * When we cast the equations in first-order flux form
```

File: src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp]: Give a pointer to what beta_background might be.  Guessing, perhaps something like `... where \f$beta_\mathrm{background}\f$ is generally \f$\Omega \ times \vec{r} + dota \vec{r} \f$`

File: src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp]:
```suggestion
 * The equations simplify significantly if the conformal metric is flat and Cartesian 
  * coordinates are used ("conformal flatness"). In this case you can specify
```

File: src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp]: Well, I wonder whether all of this is overdoing templates.  But I trust you and other spectre developers have discussed the level of templating already more than enough.

File: src/Elliptic/Systems/Xcts/Geometry.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/Geometry.hpp]:
```suggestion
}  // namespace Xcts
```

File: src/Elliptic/Systems/Xcts/Geometry.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/Geometry.hpp]: Is it smart to have a separate Xcts::Geometry that duplcaites Poisson::Geometry?   After all, the two must be used and modified always in lockstep.

File: src/Elliptic/Systems/Xcts/Tags.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/Tags.hpp]:
```suggestion
 * metric \f$\bar{\gamma}_{ij}\f$, the trace of the extrinsic curvature \f$K\f$, their
```

File: src/Elliptic/Systems/Xcts/Tags.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/Tags.hpp]:
```suggestion
 * \f$\beta^i\f$. The remaining quantities in the equations, i.e. the conformal
```

File: src/Elliptic/Systems/Xcts/Tags.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/Tags.hpp]: elsewhere I saw a comment that u_{ij} and beta_background are combined into one field.  Mention here (or somewhat further down)

File: src/Elliptic/Systems/Xcts/Tags.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/Tags.hpp]:
```suggestion
 * scenario at hand. Of particular importance is the conformal metric, which
 * defines the background geometry, the covariant derivative \f$\bar{D}\f$, 
  * the Ricci scalar \f$\bar{R}\f$ and the longitudinal operator
```

File: src/Elliptic/Systems/Xcts/Tags.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/Tags.hpp]: One difference for the shift equation is that in the GR-shift equation, the trace of the gradient is removed (i.e. longitudinal operator vs. strain).   State explicitly whether this matters, and if so how.

File: src/Elliptic/Systems/Xcts/Tags.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/Tags.hpp]:
```suggestion
 * \brief The shift vector contracted with the gradient of the trace 
 * of the extrinsic curvature: \f$\beta^i\partial_i K\f$
```

File: src/PointwiseFunctions/GeneralRelativity/Tags.hpp
Comment (haraldp271) [inline src/PointwiseFunctions/GeneralRelativity/Tags.hpp]: @markscheel @kidder Is this the most sensible name for Gamma_i, or is there already a different standardized name in the KST papers, SpEC, or elsewhere?

Comment (haraldp271) [reply src/PointwiseFunctions/GeneralRelativity/Tags.hpp]: Since the Christoffel symbols of the second kind are the standard ones nearly always used, I would suggest to just call them 'Christoffel' for brevity.   In the fewer cases where Gamma_{ijk} is needed, those can be called 'ChristoffelFirstKind'.

**Review by haraldp271 — COMMENTED** [review]
Some early feedback, mostly on comments.  
- this PR still has a lot of code from Poisson and Elasticity, which I believe is part of a separate PR, so I didn't comment on it here; will this Xcts PR be reposed in smaller form once Poisson i s merged?
- I also did NOT check the algebra or the equations.  Happy to do so next time, let me know.
- I also did NOT check the algebra in the tests -- are people worried about the same algebraic error being present in both the tests and the real code (after all, this is based  on the same calc of Nils...)

**Comment by moxcodes** [comment]
All of the PR's mentioned here appear to have been merged -- @nilsleiffischer is this ready to be rebased and reviewed as non-dependent?

**Comment by nilsvu** [comment]
It is, and I'll get this rebased ASAP

**Comment by nilsvu** [comment]
Factored out a few more PRs and updated the list of dependencies in the PR description :)

File: src/Elliptic/Systems/Elasticity/Equations.cpp
Comment (nilsvu) [reply src/Elliptic/Systems/Elasticity/Equations.cpp]: This code was merged in #2674. I think it's better to limit this PR to the XCTS equations now. I'm happy to go over the elasticity code in a separate PR to fix these kinds of things though.

File: src/Elliptic/Systems/Poisson/Equations.cpp
Comment (nilsvu) [reply src/Elliptic/Systems/Poisson/Equations.cpp]: Our tensors have two kinds of `get` functions: (1) Member functions like `tensor.get(i, j)` for runtime-access to the components and (2) free functions like `get<0, 1>(tensor)` for when the components are known at compile-time. Since a `Scalar` only has a single component we can always use the compile-time version `get(scalar)`.

File: src/Elliptic/Systems/Poisson/Equations.hpp
Comment (nilsvu) [reply src/Elliptic/Systems/Poisson/Equations.hpp]: This code was also merged in #2674

File: src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp
Comment (nilsvu) [reply src/Elliptic/Systems/Xcts/FirstOrderSystem.hpp]: This example is already in the documentation of the `ShiftExcess` tag, so I added a reference to that documentation.

File: src/Elliptic/Systems/Xcts/Geometry.hpp
Comment (nilsvu) [reply src/Elliptic/Systems/Xcts/Geometry.hpp]: Not sure, the actual code duplication here is minimal since the enums mostly exist for documentation and to write `FlatCartesian`/`Curved` instead of booleans. I think it might be better to keep the enum and its documentation specific to the system instead of trying to generalize it too much. Either could work though 🤷‍♂️

File: src/Elliptic/Systems/Xcts/Tags.hpp
Comment (nilsvu) [reply src/Elliptic/Systems/Xcts/Tags.hpp]: The longitudinal operator is actually just analogous to a particular elastic constitutive relation, I added a few notes to the docs.

**Comment by nilsvu** [comment]
@haraldp271 I have rebased this PR on latest develop and added a fixup commit to address your comments. I think this is ready for review now. Please take a look when you get the chance!

<s>In your review you can ignore the first two commits, since they have their own PRs (#2875 and #2876) and will disappear once those are merged.</s> Those two PRs are merged now, but I factored a bit more code out into to make this PR smaller. #2915 just adds the tags and some documentation.

**Comment by nilsvu** [comment]
@haraldp271 I stripped this PR down _a lot_ so it only adds the actual math. The next PR will add the code that ties everything together. Please continue your review here when you get the chance!

File: src/Elliptic/Systems/Xcts/Equations.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/Equations.hpp]: The index-placement of the tensors is surprising.  Perhaps add a _flat to the function name?

File: src/Elliptic/Systems/Xcts/Equations.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/Equations.hpp:40]: The matter term is problematic, because the unscaled rho.  See https://arxiv.org/abs/gr-qc/0610120, discussion after Eq. (2), and in fact the entire paper, which is about those problems.

Recommend you open an issue to improve on the rho scaling in the future.  E.g. code as psi^{5-n} \tilde{\rho},  where \tilde{\rho}=\psi^n \rho is now the freely specifiable data, and let the user specify the value for n.  (n=8 was suggested by York, as it preseves a certain energy condition;  n=6 might also be useful, as it preserves the total energy density \sqrt{g}rho  =  \sqrt{tilde{g}} \tilde{rho} )

File: src/Elliptic/Systems/Xcts/Equations.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/Equations.hpp:79]: Perhaps this is handled elsewhere (and I'll discover as I go on): 
A_{ij} contains the shift, so should there also be  shift_correction?

File: src/Elliptic/Systems/Xcts/Equations.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/Equations.hpp:200]: Elasticity?

File: src/Elliptic/Systems/Xcts/Equations.cpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/Equations.cpp]: four nested loops, with three multiplies in the innermost loop gives 3^4*3 = 243 multiplies.

If one had first raised an index,  $T^i_j = g^{ik} T_{kj}$  (9*3 = 27 multiplies), and then contracted   $T^i_j\,T^j_i$   (9 multiplies), then only 36 multiplies in total would be needed, but also a temporary. 

At some point it would be interesting to time where the cross-over is between temporary vs. more floating point operations.

File: src/Elliptic/Systems/Xcts/Equations.cpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/Equations.cpp:85]: SpEC pre-computes terms in the linearization.  For this specific example, SpEC would compute and store   A = (5/12 K^2 + 10*pi*rho)*psi^4.    This is done *once* per non-linear Newton-Raphson iteration.

The linearization is then evaluated as   += A*conformal_factor_correction.    This (cheap) evaluation is then evaluated at every linear iteration.

Would be interesting at which point this becomes advantageous.  Clearly, this will depend on the number of linear iterations typically taken, and the relative cost of evaluating the linearization vs. applying the preconditioner.

File: src/Elliptic/Systems/Xcts/Equations.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/Equations.hpp:67]: Please add formula defining longitudinal_shift_minus_dt_conformal_metric_over_lapse_square

File: src/Elliptic/Systems/Xcts/Equations.cpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/Equations.cpp:223]: The formulas in this function (and some others) are complicated enough that references to concrete equation numbers of your favourite paper/book would be useful.  Please add something like 
"This function implements Eqs. (xx) of Baumgarte+Shapiro "Numerical Relativity"  at the relevant places.

**Review by haraldp271 — DISMISSED** [review]
My comments are trivial enough that I trust Nils to implement them before merging, w/o me seeing this again.

File: src/Elliptic/Systems/Xcts/Equations.hpp
Comment (nilsvu) [reply src/Elliptic/Systems/Xcts/Equations.hpp:200]: Yep, the elasticity function can be used here as-is. It just adds the Christoffel-terms for the divergence of a symmetric rank-two tensor field (the elastic stress / longitudinal shift)

File: src/Elliptic/Systems/Xcts/Equations.cpp
Comment (nilsvu) [reply src/Elliptic/Systems/Xcts/Equations.cpp]: I actually have some temporary memory to spare at the points where this function is used, so I refactored it to perform 63 multiplications instead of the 243 (63 because the two tensors may not be the same) without allocating more memory.

File: src/Elliptic/Systems/Xcts/Equations.hpp
Comment (nilsvu) [reply src/Elliptic/Systems/Xcts/Equations.hpp:40]: I just went ahead and added the conformal scaling of the matter terms. I pushed the change as a fixup commit because it's not entirely trivial.

File: src/Elliptic/Systems/Xcts/Equations.cpp
Comment (nilsvu) [reply src/Elliptic/Systems/Xcts/Equations.cpp:85]: Opened https://github.com/sxs-collaboration/spectre/issues/3071 for this. Pre-computing these quantities could potentially speed up current solves because they explicitly build the subdomain operator by evaluating the linearization many times, but the performance gain might decrease with better preconditioners.

File: src/Elliptic/Systems/Xcts/Equations.hpp
Comment (nilsvu) [reply src/Elliptic/Systems/Xcts/Equations.hpp:67]: Added references to the tag that defines this quantity (and `longitudinal_shift_minus_dt_conformal_metric_square` as well).

**Comment by nilsvu** [comment]
Thanks for reviewing @haraldp271! I pushed a few fixup commits since they were not entirely non-trivial. Please take a look at the changes when you get the chance.

File: src/Elliptic/Systems/Xcts/Geometry.hpp
Comment (haraldp271) [inline src/Elliptic/Systems/Xcts/Geometry.hpp:8]: I don't understand why we need separate, identical tags in Poisson and Xcts.  Wouldn't code re-use point toward just having one Geometry tag?   And also, are separate Poisson::Geometry and Xcts::Geometry tags going to cause inconsistencies later on?  E.g. what if Geometry-templated XCTS code wants to re-use Poisson-code?

**Review by haraldp271 — DISMISSED** [review]

File: src/Elliptic/Systems/Xcts/Equations.hpp
Comment (nilsvu) [reply src/Elliptic/Systems/Xcts/Equations.hpp:79]: You probably discovered later that the shift correction is handled by `add_*_momentum_sources`. This is so the functions can also be used when not solving for the shift.

File: src/Elliptic/Systems/Xcts/Geometry.hpp
Comment (nilsvu) [reply src/Elliptic/Systems/Xcts/Geometry.hpp:8]: The enums could be merged, though I didn't do that so far because:

1. I don't think the code re-use is super important here since it's just an enum to get human-readable labels and there's no logic involved. Generalizing and re-using code is nice, but taken too far it can make the code more difficult to follow.
2. I'm not sure where to place it in the codebase since `elliptic::Geometry` doesn't seem right; it could be more general than the elliptic code, but such generality is probably unnecessary.
3. I don't think inconsistencies with Poisson code will be an issue. The Poisson (and elasticity) code is called internally to re-use implementations, but none of that is exposed to the public interface. Also the Poisson and elasticity functions are called directly (e.g. `Poisson::curved_fluxes(...)`), not using the `Poisson::Geometry` enum. The enum is there just to have nicer labels for the template parameters.

**Comment by nilsvu** [comment]
@nilsdeppe would you be willing to give this a second review?

**Comment by nilsdeppe** [comment]
Sure!

File: src/Elliptic/Systems/Xcts/Equations.hpp
Comment (nilsdeppe) [inline src/Elliptic/Systems/Xcts/Equations.hpp]: could you please document what `w_j` and `v_j` are?

File: src/Elliptic/Systems/Xcts/Equations.cpp
Comment (nilsdeppe) [inline src/Elliptic/Systems/Xcts/Equations.cpp:94]: [optional] I have no objection to having numbers like `0.03125` here, but it would be nice to have a comment of what fraction it is :) If you decide to add a comment for this here, could you also do so in the below functions?

**Review by nilsdeppe — DISMISSED** [review]
LGTM, couple small documentation requests. Please squash if you agree with the suggestions :)

File: src/Elliptic/Systems/Xcts/Equations.hpp
Comment (nilsvu) [reply src/Elliptic/Systems/Xcts/Equations.hpp]: They are defined in the docs for the XCTS system which isn't in this PR yet, so I replaced them with `\partial_j (\alpha\psi)` and `\partial_j \psi`.

File: src/Elliptic/Systems/Xcts/Equations.cpp
Comment (nilsvu) [reply src/Elliptic/Systems/Xcts/Equations.cpp:94]: done 👍

**Comment by nilsvu** [comment]
Squashed in the added comments and rebased. Thanks for reviewing @haraldp271 and @nilsdeppe!

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsvu** [comment]
@haraldp271 Could you sign off on this before we merge it? :)

File: src/Elliptic/Systems/Xcts/Geometry.hpp
Comment (haraldp271) [reply src/Elliptic/Systems/Xcts/Geometry.hpp:8]: I am fine if you're fine.

**Review by haraldp271 — APPROVED** [review]
my comments are taken care of.  So happy to have this merged.

**Pull request merged**
